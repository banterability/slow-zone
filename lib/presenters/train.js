// Generated by CoffeeScript 2.2.1
(function() {
  var Dateline, STRING_FIELDS, Train, getNativeBoolean, getNativeDate, getNativeFloat, getNativeInteger;

  Dateline = require('dateline');

  STRING_FIELDS = {
    'destNm': 'destinationName',
    'rt': 'routeId',
    'staNm': 'stationName',
    'stpDe': 'stopDescription'
  };

  Train = class Train {
    constructor(attributes = {}) {
      var attributeName, methodName;
      this.attributes = attributes;
// Map string fields that don't need additional processing
      for (attributeName in STRING_FIELDS) {
        methodName = STRING_FIELDS[attributeName];
        this[methodName] = (function(value) {
          return function() {
            return value;
          };
        })(this.attributes[attributeName]);
      }
    }

    //# Boolean Fields
    isApproaching() {
      return getNativeBoolean(this.attributes.isApp);
    }

    isDelayed() {
      return getNativeBoolean(this.attributes.isDly);
    }

    isFaulty() {
      return getNativeBoolean(this.attributes.isFlt);
    }

    isScheduled() {
      return getNativeBoolean(this.attributes.isSch);
    }

    //# Date Fields
    arrivalTime() {
      return getNativeDate(this.attributes.arrT);
    }

    predictionTime() {
      return getNativeDate(this.attributes.prdt);
    }

    //# Floating Point Fields
    latitude() {
      return getNativeFloat(this.attributes.lat);
    }

    longitude() {
      return getNativeFloat(this.attributes.lon);
    }

    //# Generated Fields
    arrivalMinutes() {
      return Math.round((this.arrivalTime() - this.predictionTime()) / (60 * 1000));
    }

    arrivalString() {
      return Dateline(this.arrivalTime()).getAPTime();
    }

    predictionAge() {
      return Math.round((new Date() - this.predictionTime()) / 1000);
    }

    route() {
      var route;
      switch (route = this.routeId()) {
        case "Brn":
          return "Brown";
        case "G":
          return "Green";
        case "Org":
          return "Orange";
        case "P":
          return "Purple";
        case "Y":
          return "Yellow";
        default:
          return route;
      }
    }

    routeClass() {
      return this.route().toLowerCase();
    }

    //# Integer Fields
    destinationId() {
      return getNativeInteger(this.attributes.destSt);
    }

    directionId() {
      return getNativeInteger(this.attributes.trDr);
    }

    heading() {
      return getNativeInteger(this.attributes.heading);
    }

    runNumber() {
      return getNativeInteger(this.attributes.rn);
    }

    stationId() {
      return getNativeInteger(this.attributes.staId);
    }

    stopId() {
      return getNativeInteger(this.attributes.stpId);
    }

    toHash() {
      return {
        destination: {
          id: this.destinationId(),
          name: this.destinationName()
        },
        location: {
          latitude: this.latitude(),
          longitude: this.longitude(),
          heading: this.heading()
        },
        prediction: {
          arrivalMinutes: this.arrivalMinutes(),
          arrivalString: this.arrivalString(),
          arrivalTime: this.arrivalTime(),
          predictionAge: this.predictionAge(),
          predictionTime: this.predictionTime()
        },
        route: {
          class: this.routeClass(),
          directionId: this.directionId(),
          id: this.routeId(),
          name: this.route(),
          run: this.runNumber()
        },
        station: {
          id: this.stationId(),
          name: this.stationName(),
          stop: {
            id: this.stopId(),
            description: this.stopDescription()
          }
        },
        status: {
          approaching: this.isApproaching(),
          delayed: this.isDelayed(),
          faulty: this.isFaulty(),
          scheduled: this.isScheduled()
        }
      };
    }

  };

  module.exports = Train;

  //# Helpers
  getNativeBoolean = function(booleanString) {
    return getNativeInteger(booleanString) === 1;
  };

  getNativeDate = function(timeString) {
    var day, hour, min, month, sec, str, year;
    [str, year, month, day, hour, min, sec] = timeString.match(/(\d{4})(\d{2})(\d{2}) (\d{2}):(\d{2}):(\d{2})/);
    return new Date(year, month - 1, day, hour, min, sec);
  };

  getNativeFloat = function(floatString) {
    return parseFloat(floatString);
  };

  getNativeInteger = function(integerString) {
    return parseInt(integerString, 10);
  };

}).call(this);
