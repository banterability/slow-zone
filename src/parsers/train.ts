import { Destination } from "./destination";
import { Location } from "./location";
import { Prediction } from "./prediction";
import { Route } from "./route";
import { Station } from "./station";
import { Status } from "./status";
import type { TrainResponse } from "../types/responses";

export class Train {
  attributes: TrainResponse;

  constructor(attributes) {
    this.attributes = attributes;
  }

  toJSON() {
    return {
      destination: new Destination(this.attributes).toJSON(),
      location: new Location(this.attributes).toJSON(),
      prediction: new Prediction(this.attributes).toJSON(),
      route: new Route(this.attributes).toJSON(),
      station: new Station(this.attributes).toJSON(),
      status: new Status(this.attributes).toJSON(),
    };
  }
}
