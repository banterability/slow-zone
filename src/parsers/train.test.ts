const timekeeper = require("timekeeper");
import { parseTrain } from "./train";

const MOCK_TRAIN_JSON = JSON.stringify({
  staId: "40160",
  stpId: "30031",
  staNm: "LaSalle/Van Buren",
  stpDe: "Service at Inner Loop platform",
  rn: "715",
  rt: "Org",
  destSt: "30182",
  destNm: "Midway",
  trDr: "5",
  prdt: "2014-10-07T14:49:27",
  arrT: "2014-10-07T14:50:27",
  isApp: "1",
  isSch: "0",
  isDly: "0",
  isFlt: "0",
  flags: "",
  lat: "41.87685",
  lon: "-87.6327",
  heading: "269",
});

describe("Train", () => {
  describe("toJSON", () => {
    beforeAll(() => {
      timekeeper.freeze(new Date(2014, 9, 7, 14, 50, 57));
    });

    afterAll(() => {
      timekeeper.reset();
    });

    test("returns all attributes", () => {
      const train = parseTrain(JSON.parse(MOCK_TRAIN_JSON));

      expect(train).toMatchSnapshot();
    });
  });
});
