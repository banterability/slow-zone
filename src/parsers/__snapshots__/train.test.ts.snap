// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Train > toJSON > returns all attributes 1`] = `
{
  "destination": {
    "id": 30182,
    "name": "Midway",
  },
  "location": {
    "heading": 269,
    "latitude": 41.87685,
    "longitude": -87.6327,
  },
  "prediction": {
    "arrivalMinutes": 1,
    "arrivalString": "2:50 PM",
    "arrivalTime": 2014-10-07T19:50:27.000Z,
    "predictionAge": 60,
    "predictionTime": 2014-10-07T19:49:27.000Z,
  },
  "route": {
    "class": "orange",
    "directionId": 5,
    "id": "Org",
    "name": "Orange",
    "run": 715,
  },
  "station": {
    "id": 40160,
    "name": "LaSalle/Van Buren",
    "stop": {
      "description": "Service at Inner Loop platform",
      "id": 30031,
    },
  },
  "status": {
    "approaching": true,
    "delayed": false,
    "faulty": false,
    "scheduled": false,
  },
}
`;
